lib/ansible/modules/database/postgresql/postgresql_user.py

```diff
@@ -273,41 +286,72 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
+        # on some databases, E.g. AWS RDS instances, there is no access to
+        # the pg_authid relation to check the pre-existing password, so we
+        # just assume password is different
+        return True
+
+    # Do we actually need to do anything?
+
+    if encrypted == 'ENCRYPTED' and password.startswith('md5'):
+        return password != current_role_attrs['rolpassword']
+
+    if encrypted == 'ENCRYPTED' and not password.startswith('md5'):
+        if passlib_hash_found:
+            return pm.encrypt(password, user) != current_role_attrs['rolpassword']
+        else:
+            # Can't check the password, so just assume update needed.  A warning would be nice.
+            return True
+
+    return password != current_role_attrs['rolpassword']
+
+
+def user_alter(db_connection, module, user, password, role_attr_flags, encrypted, expires, no_password_changes):
     """Change user password and/or attributes. Return True if changed, False otherwise."""
     changed = False
 
-    # Note: role_attr_flags escaped by parse_role_attrs and encrypted is a literal
+    cursor = db_connection.cursor(cursor_factory=psycopg2.extras.DictCursor)
+    # Note: role_attr_flags escaped by parse_role_attrs and encrypted is a
+    # literal
     if user == 'PUBLIC':
         if password is not None:
             module.fail_json(msg="cannot change the password for PUBLIC user")
         elif role_attr_flags != '':
-            module.fail_json(msg="cannot change the role_attr_flags for PUBLIC user")
+            module.fail_json(
+                msg="cannot change the role_attr_flags for PUBLIC user")
         else:
             return False
 
     # Handle passwords.
     if not no_password_changes and (password is not None or role_attr_flags != '' or expires is not None):
         # Select password and all flag-like columns in order to verify changes.
-        select = "SELECT * FROM pg_authid where rolname=%(user)s"
-        cursor.execute(select, {"user": user})
-        # Grab current role attributes.
-        current_role_attrs = cursor.fetchone()
+        query_password_data = dict(password=password, expires=expires)
 
-        # Do we actually need to do anything?
-        pwchanging = False
-        if password is not None:
-            # 32: MD5 hashes are represented as a sequence of 32 hexadecimal digits
-            #  3: The size of the 'md5' prefix
-            # When the provided password looks like a MD5-hash, value of
-            # 'encrypted' is ignored.
-            if ((password.startswith('md5') and len(password) == 32 + 3) or encrypted == 'UNENCRYPTED'):
-                if password != current_role_attrs['rolpassword']:
-                    pwchanging = True
-            elif encrypted == 'ENCRYPTED':
-                hashed_password = 'md5{0}'.format(md5(to_bytes(password) + to_bytes(user)).hexdigest())
-                if hashed_password != current_role_attrs['rolpassword']:
-                    pwchanging = True
```

Comment:
discussion with @pilou- move this part into the "should we change password" function above